tap_typeform/streams.py

```diff
@@ -292,76 +318,52 @@ def sync_forms(atx):
-
-            ut_current_date = int(current_date.timestamp())
-            LOGGER.info('ut_current_date: {} '.format(ut_current_date))
-            ut_next_date = int(next_date.timestamp())
-            LOGGER.info('ut_next_date: {} '.format(ut_next_date))
-            [responses, max_submitted_at] = sync_form(atx, form_id, ut_current_date, ut_next_date)
-            # if the max responses were returned, we have to make the call again
-            # going to increment the max_submitted_at by 1 second so we don't get dupes,
-            # but this also might cause some minor loss of data.
-            # there's no ideal scenario here since the API has no other way than using
-            # time ranges to step through data.
-            bookmark_value = construct_bookmark(stream_name, next_date)
-
-            while responses == 1000:
-                interim_next_date = pendulum.parse(max_submitted_at) + datetime.timedelta(seconds=1)
-                ut_interim_next_date = int(interim_next_date.timestamp())
-                bookmark_value = construct_bookmark(stream_name, interim_next_date)
-                write_forms_state(atx, stream_name, form_id, bookmark_value)
-                [responses, max_submitted_at] = sync_form(atx, form_id, ut_interim_next_date, ut_next_date)
-
-            # check if next_date is in the future
-            now_parsed = pendulum.parse(now.isoformat())
-            next_date = now_parsed if next_date > now else next_date
-
-            # if the prior sync is successful it will write the date_to_resume bookmark
+        page_count = 0
+
+        ut_current_date = int(current_date.timestamp())
+        LOGGER.info('ut_current_date: {} '.format(ut_current_date))
+        
+        page_count, max_submitted_at, max_token = sync_form(atx, form_id, ut_current_date)
+
+        parsed_max_submitted_at = pendulum.parse(max_submitted_at)
+        bookmark_value = construct_bookmark(stream_name, parsed_max_submitted_at)
+
+        # if the max responses were returned, we have to make the call again
+        # using the max_token from the initial call.
+        while page_count > 1:
+            page_count, max_submitted_at, max_token = sync_form(atx,
+                                                                form_id,
+                                                                ut_current_date,
+                                                                token=max_token,
+                                                                next_page=True)
+
+            parsed_max_submitted_at = pendulum.parse(max_submitted_at)
+            bookmark_value = construct_bookmark(stream_name, parsed_max_submitted_at)
             write_forms_state(atx, stream_name, form_id, bookmark_value)
-            current_date = next_date
 
-    if 'forms'in atx.selected_stream_ids:
+        # check if bookmark is in the past
+        now_parsed = pendulum.parse(now.isoformat())
```

Comment:
`pendulum.parse()` returns a `pendulum.datetime.DateTime` object.

```python
>>> type(pendulum.parse("2021-09-22T12:34:56Z"))
<class 'pendulum.datetime.DateTime'>
```

Why do we need to call `parse()` if `now` is a `datetime.datetime` object?

Reply:
This is due to the `construct_bookmark` function using a pendulum only method called `to_datetime_string()`. But I can rewrite it with the `isoformat()` method from datetime if that will be cleaner. 

Will need to test it doesn't break other parts of the tap that use that function.

Reply:
Sorry, I should've just typed out my question and my reasoning for asking.

For the sake of comparing date time objects, we don't care if they are `pendulum` date times or `datetime` date times. 

```python
>>> import pendulum
>>> from datetime import datetime
>>> import pytz
>>> now = datetime.now(pytz.utc)
>>> old_pendulum_time = pendulum.parse("2021-09-01T12:34:56Z")
>>> new_pendulum_time = pendulum.parse("2021-12-31T12:34:56Z")

>>> now
datetime.datetime(2021, 9, 24, 14, 28, 44, 430580, tzinfo=<UTC>)
>>> old_pendulum_time
DateTime(2021, 9, 1, 12, 34, 56, tzinfo=Timezone('UTC'))
>>> new_pendulum_time
DateTime(2021, 12, 31, 12, 34, 56, tzinfo=Timezone('UTC'))

# Can we compare `datetime.datetime` to `pendulum.datetime`?
>>> max(now, old_pendulum_time) == now
True
# Does order matter?
>>> max(old_pendulum_time, now) == now
True

>>> max(now, new_pendulum_time) == new_pendulum_time
True
>>> max(new_pendulum_time, now) == new_pendulum_time
True

>>> max(old_pendulum_time, new_pendulum_time, now) == new_pendulum_time
True
```

So that's why I was questioning `now_parsed = pendulum.parse(now.isoformat())`. It seems like changing

```diff
- now = datetime.datetime.now(pytz.utc)
- s_d = now.replace(hour=0, minute=0, second=0, microsecond=0)
+ now_parsed = datetime.datetime.now(pytz.utc)
+ s_d = now_parsed.replace(hour=0, minute=0, second=0, microsecond=0)
```

would be more consistent with your variable naming (`foo` is a datetime string, and `foo_parsed` is a datetime object) and it gets rid of this `now_parsed = pendulum.parse(now.isoformat())` line altogether

Reply:
Looking at this code a little closer, what do we gain from moving the bookmark out of the past? Correct me if I'm wrong, but I think the `write_forms_state()` on the last page of results puts you in more or less the same state. On top of that, it removes this final comparison against the time at the start of the sync

Reply:
Agreed on the pendulum vs datatime issues. I used the datetime module as that's what I'm familiar with, but removing it makes more sense to me since the rest of the tap is using pendulum and there won't be a need to convert between the two.

Reply:
The reason for moving the bookmark forward is due to how the tap works for syncing the `answers` stream and the bookmark integration test. I had initially removed that part, but the bookmark test kept failing. This is because it compares the bookmarks for the whole stream, but the tap is actually setting multiple bookmarks (per form) on that stream. This is how the tap was originally implemented. Tap tester output below with the comparison removed, one of the forms `iRnSCYFE` has a `landed_at` that is in the past. Since we're requesting all responses to the form setting the bookmark to `now_parsed` will not miss any responses on the next sync.

```bash
main - new state: {"bookmarks": {"answers": {"iRnSCYFE": {"landed_at": "2021-08-23 18:57:24"}, "SDKWHq1e": {"landed_at": "2021-09-22 23:44:23"}, "Tk5DoxcG": {"landed_at": "2021-09-22 23:43:54"}}, "forms": {"last_updated_at": "2021-07-07T19:58:37.543354Z"}}}
         target - INFO forms (2): Batch is valid
         target - INFO Requests complete, stopping loop
setting state to {'bookmarks': {'answers': {'iRnSCYFE': {'landed_at': '2021-08-23 18:57:24'}, 'SDKWHq1e': {'landed_at': '2021-09-22 23:44:23'}, 'Tk5DoxcG': {'landed_at': '2021-09-22 23:43:54'}}, 'forms': {'last_updated_at': '2021-07-07T19:58:37.543354Z'}}}
**************************************************************************************************************
sync mode exitted succesfully
OK: tap_exit_status for sync job(5-1632509227) is 0
OK: target_exit_status for job(5-1632509227) is 0
OK: discovery_exit_status for job(5-1632509227) is 0
OK: check_exit_status for job(5-1632509227) is None
Output records passing schema validation:
        landings = 15   pks: {'landing_id'}
        answers = 30    pks: {'landing_id', 'question_id'}
        questions = 7   pks: {'form_id', 'question_id'}
        forms = 2       pks: {'id'}
total replicated row count: 54
.
======================================================================
FAIL: test_run (test_typeform_bookmarks.TypeformBookmarks) (stream='answers')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/code/tap-typeform/tests/test_typeform_bookmarks.py", line 189, in test_run
    self.assertLessEqual(
AssertionError: '2021-09-22T23:44:12Z' not less than or equal to '2021-08-23T18:57:24Z' : Second sync bookmark was set incorrectly, a record with a greater replication-key value was synced.

----------------------------------------------------------------------
Ran 4 tests in 18.311s

FAILED (failures=1)
```

The tap-tester output with those lines added back:
```bash
 main - new state: {"bookmarks": {"answers": {"iRnSCYFE": {"landed_at": "2021-09-24 18:55:43"}, "SDKWHq1e": {"landed_at": "2021-09-24 18:55:43"}, "Tk5DoxcG": {"landed_at": "2021-09-24 18:55:44"}}, "forms": {"last_updated_at": "2021-07-07T19:58:37.543354Z"}}}
         target - INFO Requests complete, stopping loop
setting state to {'bookmarks': {'answers': {'iRnSCYFE': {'landed_at': '2021-09-24 18:55:43'}, 'SDKWHq1e': {'landed_at': '2021-09-24 18:55:43'}, 'Tk5DoxcG': {'landed_at': '2021-09-24 18:55:44'}}, 'forms': {'last_updated_at': '2021-07-07T19:58:37.543354Z'}}}
**************************************************************************************************************
sync mode exitted succesfully
OK: tap_exit_status for sync job(5-1632509744) is 0
OK: target_exit_status for job(5-1632509744) is 0
OK: discovery_exit_status for job(5-1632509744) is 0
OK: check_exit_status for job(5-1632509744) is None
Output records passing schema validation:
        landings = 15   pks: {'landing_id'}
        answers = 30    pks: {'question_id', 'landing_id'}
        questions = 7   pks: {'form_id', 'question_id'}
        forms = 2       pks: {'id'}
total replicated row count: 54
.
----------------------------------------------------------------------
Ran 4 tests in 18.272s

OK
```

Reply:
I was not advocating for removing `datetime`. It's also the module I'm more familiar with.

I was just saying that converting from`datetime` to `pendulum` didn't need to happen since Python can compare the two objects.

Reply:
> Since we're requesting all responses to the form setting the bookmark to now_parsed will not miss any responses on the next sync.

I agree with this

Reply:
> the bookmark test kept failing. This is because it compares the bookmarks for the whole stream, but the tap is actually setting multiple bookmarks (per form) on that stream.

Let me check with the QA team. I agree that this test is traditionally suited for streams that write one bookmark. 

But we should be able to assert against each form since each form appears in the state